packages/context/src/resolver.ts

```diff
@@ -20,6 +26,151 @@ export type Constructor<T> =
+   * @param binding Binding
+   * @param session Resolution session
+   */
+  static enterBinding(
+    binding: Binding,
+    session?: ResolutionSession,
+  ): ResolutionSession {
+    session = session || new ResolutionSession();
+    session.enter(binding);
+    return session;
+  }
+
+  /**
+   * Push an injection into the session
+   * @param injection Injection
+   * @param session Resolution session
+   */
+  static enterInjection(
+    injection: Injection,
+    session?: ResolutionSession,
+  ): ResolutionSession {
+    session = session || new ResolutionSession();
+    session.enterInjection(injection);
+    return session;
+  }
+
+  private describeInjection(injection?: Injection) {
+    if (injection == null) return injection;
+    const name = DecoratorFactory.getTargetName(
+      injection.target,
+      injection.member,
+      injection.methodDescriptorOrParameterIndex,
+    );
+    return {
+      targetName: name,
+      bindingKey: injection.bindingKey,
+      metadata: injection.metadata,
+    };
+  }
+
+  /**
+   * Push the injection onto the session
+   * @param injection Injection
+   */
+  enterInjection(injection: Injection) {
+    if (debug.enabled) {
+      debug('Enter injection:', this.describeInjection(injection));
+    }
+    this.injections.push(injection);
+    if (debug.enabled) {
+      debug('Injection path:', this.getInjectionPath());
+    }
+  }
+
+  /**
+   * Pop the last injection
+   */
+  exitInjection() {
+    const injection = this.injections.pop();
+    if (debug.enabled) {
+      debug('Exit injection:', this.describeInjection(injection));
+      debug('Injection path:', this.getInjectionPath() || '<empty>');
+    }
+    return injection;
+  }
+
+  /**
+   * Getter for the current binding
+   */
+  get injection() {
+    return this.injections[this.injections.length - 1];
+  }
+
+  /**
+   * Getter for the current binding
+   */
+  get binding() {
+    return this.bindings[this.bindings.length - 1];
+  }
+
+  /**
+   * Enter the resolution of the given binding. If
+   * @param binding Binding
+   */
+  enter(binding: Binding) {
+    if (debug.enabled) {
+      debug('Enter binding:', binding.toJSON());
+    }
+    if (this.bindings.indexOf(binding) !== -1) {
+      throw new Error(
+        `Circular dependency detected for '${
+          binding.key
+        }' on path '${this.getBindingPath()}'`,
+      );
+    }
+    this.bindings.push(binding);
+    if (debug.enabled) {
+      debug('Binding path:', this.getBindingPath());
+    }
+  }
+
+  /**
+   * Exit the resolution of a binding
+   */
+  exit() {
+    const binding = this.bindings.pop();
+    if (debug.enabled) {
+      debug('Exit binding:', binding && binding.toJSON());
+      debug('Binding path:', this.getBindingPath() || '<empty>');
+    }
+    return binding;
+  }
+
+  /**
+   * Get the binding path as `bindingA->bindingB->bindingC`.
+   */
+  getBindingPath() {
+    return this.bindings.map(b => b.key).join('->');
```

Comment:
Love this idea, but could we change the join formatting to ` --> ` ? Just to provide some space for readability, IMO.

Reply:
Which one do you prefer (`_` represents a space)?
1. `_->_`
2. `-->`
3. `_-->_`

Reply:
3 would be the way to go, I think. (`bindingA --> bindingB --> bindingC`)

Reply:
ok, fixed